### PR TITLE
Subnetwork and Last Layer Laplace Models

### PR DESCRIPTION
**Pull request by MarkArdman**
This pull request introduces the implementation of Subnetwork and Last Layer Laplace Models. They can be utilized by passing arg subset_of_weights as :last_layer or :subnetwork. Subnetwork model also requires a subnetwork_indices argument. For user convenience, it is represented as an array of coordinates.

### CODE REVIEW

**Comment by adelinacazacu** [comment]
This seems to work, but I believe we are supposed to select the weights differently. From what I spoke with Severin, your implementation only gives the user the option to select full layers for the subset of weights, but it should let you select fragments of the layers as well. So if you think of your list of weights and biases as a list, you should be able to select completely random subsets. You can visualize this better on page 4, fig 1 of the Laplace Redux paper.
I think this is where it gets complicated, I'm currently trying to figure out a way we can do this. Any ideas?

File: src/baselaplace.jl [hunk 1]
```diff
@@ -35,6 +35,9 @@ function get_params(la::BaseLaplace)
         params = [θ for θ in params]                                         # get all parameters and constants in logitbinarycrossentropy
     elseif la.subset_of_weights == :last_layer
         params = [params[n_elements - 1], params[n_elements]]                # only get last layer parameters and constants
+    elseif la.subset_of_weights == :sub_network
+        indeces = la.sub_network_indices
+        params = [params[i] for i in indeces]                                # only get parameters and constants of the sub-network
     end                                                                      # params[n_elements-1] is the weight matrix of the last layer
```
[the remark this answers is not in the document]
Comment (adelinacazacu) [reply src/baselaplace.jl]: Thank you, this is very helpful insight! In this case, what do you think subnetwork_indices would look like in Julia?
And in regards to the performance advantage of this, according to the Laplace Redux paper, it should help with storing the covariance matrix for very large neural networks. However, I believe they also recognize this technique to not be the most helpful, because finding an optimal subnetwork involves combinatorial computations. So my short answer to this would be - I honestly have no clue.:)

[the remark this answers is not in the document]
Comment (severinbratus) [reply src/baselaplace.jl]: > In this case, what do you think subnetwork_indices would look like in Julia?
So at the internal level, we will represent these as an array (vector / AbstractArray) of *flattened* indices, id est integers corresponding to the flattened representation of theta (all model parameters in one vector, just like `torch.nn.utils.parameters_to_vector` does it).
Yes, PA said having a smaller J would help for faster multiplication.
> finding an optimal subnetwork involves combinatorial computations.
For now, this is something we leave to the user. There is a technique referenced in Daxberger et al. of selecting these parameters to optimize some metric, like marginal variance

File: test/laplace.jl [hunk 2]
```diff
@@ -26,6 +26,35 @@ using Statistics
...
+        subnetwork_indices=[[1, 1, 1, 1], [6, 1, 1], [4, 1]],
+    )
+    la = Laplace(
+        nn;
+        likelihood=:classification,
+        subset_of_weights=:subnetwork,
+        subnetwork_indices=[[1, 1, 1], [3, 1, 1], [4, 1]],
+    )
+    @test la.n_params == 3
+    @test la.curvature.subnetwork_indices == [1, 7, 9]
```
Comment (severinbratus) [inline test/laplace.jl:57]: This is nicely meticulous, meets the SQT standard 😎

File: test/laplace.jl [hunk 3]
```diff
@@ -180,6 +209,18 @@ end
...
+                    λ=λ,
+                    subset_of_weights=:subnetwork,
+                    subnetwork_indices=[[1, 1, 1], [2, 1], [2, 2], [3, 1, 16], [4, 1]],
+                )
+                fit!(la, data)
+                optimize_prior!(la; verbose=true)
+                plot(la, X, y)                              # standard
+                plot(la, X, y; xlims=(-5, 5), ylims=(-5, 5))  # lims
+                plot(la, X, y; link_approx=:plugin)         # plugin approximation
             end
```
Comment (severinbratus) [inline test/laplace.jl:224]: We should also add tests for the conversion procedure, though perhaps not in this MR

File: dev/notebooks/multi-class/Multi-Class-Julia-FGD.ipynb
[the remark this answers is not in the document]
Comment (MarkArdman) [reply dev/notebooks/multi-class/Multi-Class-Julia-FGD.ipynb:1]: Yeah I did move it to the notebooks folder, right? That's why as activation path I set ../../../, going out thrice, and thus ending up in the package folder.

File: src/laplace.jl [hunk 4]
```diff
@@ -48,7 +50,27 @@ function Laplace(model::Any; likelihood::Symbol, kwargs...)
...
+        @assert (args.subnetwork_indices !== nothing) "If `subset_of_weights` is `:subnetwork`, then `subnetwork_indices` should be a vector of vectors of integers."
+        params = Flux.params(model)
+        # Initialise a set of vectors 
+        selected = Set{Vector{Int}}()
+        for i in 1:length(args.subnetwork_indices)
+            @assert !(args.subnetwork_indices[i] in selected) "Element $(i) in `subnetwork_indices` should be unique."
+            @assert (args.subnetwork_indices[i][1] in 1:length(params)) "The first index of element $(i) in `subnetwork_indices` should be between 1 and $(length(params))."
+            # Calculate numnber of dimensions of a parameter 
+            n_dims = length(size(params[args.subnetwork_indices[i][1]]))
+            @assert length(args.subnetwork_indices[i]) - 1 == n_dims "Element $(i) in `subnetwork_indices` should have $(n_dims) indices."
```
[the remark this answers is not in the document]
Comment (MarkArdman) [reply src/laplace.jl:64]: Sorry, I don't quite understand what the issue is with the error message. Index i in subnetwork_indices contains coordinates + 1, meaning that it minus 1 should equal to the number of dimensions of a parameter i is pointing to.

File: src/laplace.jl [hunk 5]
```diff
@@ -76,11 +99,28 @@ function Laplace(model::Any; likelihood::Symbol, kwargs...)
...
+            likelihood,
+            params,
+            la.subset_of_weights,
+            convert_subnetwork_indices(la.subnetwork_indices, params),
+        )
+    else
+        la.curvature = getfield(Curvature, args.backend)(
+            nn, likelihood, params, la.subset_of_weights, la.subnetwork_indices
+        )
+    end
```
[the remark this answers is not in the document]
Comment (MarkArdman) [reply src/laplace.jl:117]: Gonna be real, ternary notation is also a personal favorite, accepted

File: src/laplace.jl [hunk 6]
```diff
@@ -95,7 +135,28 @@ function Laplace(model::Any; likelihood::Symbol, kwargs...)
...
+    subnetwork_indices::Vector{Vector{Int}}, params::AbstractArray
+)
+    converted_indices = Vector{Int}()
+    for i in subnetwork_indices
+        Ca = reduce((acc, p) -> acc + length(p), params[1:(i[1] - 1)]; init=0)
+        if length(i) == 2
+            push!(converted_indices, Ca + i[2])
+        elseif length(i) == 3
+            push!(converted_indices, Ca + (i[2] - 1) * size(params[i[1]], 2) + i[3])
+        end
```
[the remark this answers is not in the document]
Comment (MarkArdman) [reply src/laplace.jl:153]: I was thinking about just this today morning, which is why I generalized the error handling for higher dimensions, not sure about this though, might be worth changing as well.

File: dev/notebooks/auto-differentiation/differentiation.ipynb
[the remark this answers is not in the document]
Comment (MarkArdman) [reply dev/notebooks/auto-differentiation/differentiation.ipynb:1]: We just added another notebook and the dev folder started to feel fairly unorganized, so my OCD couldn't live with it

File: src/laplace.jl [hunk 7]
```diff
@@ -95,7 +135,28 @@ function Laplace(model::Any; likelihood::Symbol, kwargs...)
...
+
+Converts the subnetwork indices from the user given format [theta, row, column] to an Int i that corresponds to the index
+of that weight in the flattened array of weights.
+"""
+function convert_subnetwork_indices(
+    subnetwork_indices::Vector{Vector{Int}}, params::AbstractArray
+)
+    converted_indices = Vector{Int}()
+    for i in subnetwork_indices
+        Ca = reduce((acc, p) -> acc + length(p), params[1:(i[1] - 1)]; init=0)
```
[the remark this answers is not in the document]
Comment (MarkArdman) [reply src/laplace.jl:148]: Agreed

File: src/laplace.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (severinbratus) [reply src/laplace.jl:64]: Yes, I suppose this is a matter of semantics -- I interpreted the first integer (the index of the theta) as an part of the indices. Perhaps the way to make this clear to the user is:

`"Element $(i) in `subnetwork_indices` should have $(n_dims) coordinates"`

File: dev/notebooks/multi-class/Multi-Class-Julia-FGD.ipynb
[the remark this answers is not in the document]
Comment (MarkArdman) [reply dev/notebooks/multi-class/Multi-Class-Julia-FGD.ipynb:1]: Honestly, I wish I knew, I also thought it was ../.. at first

**Review by severinbratus — APPROVED** [review]

**Comment by adelinacazacu** [comment]
For the last cell in the notebook, I get this error. 
image
I see you modified the convert_subnetwork_indices() function, why was this needed?

File: src/laplace.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (Andrei32Ionescu) [reply src/laplace.jl:159]: I am also quite perplexed by this. In which case would we get `Flux.params(model)` to return a datatype of higher dimension?

[the remark this answers is not in the document]
Comment (severinbratus) [reply src/laplace.jl:159]: One example I can think of a CNN that accepts an RGB 100 pixel image of dimensions 100x100x3, then the conv. filter could be 10x10x3. However I do not know how Flux would handle such a case

**Comment by Andrei32Ionescu** [comment]
Aside from my comment above, of which I am still somewhat unsure, but which we can explore at a later date, I believe this code is well written and tested. I approve of merging it! I do suggest possibly adding some benchmarks in the `subnetworks_laplace.ipynb` notebook to see how much faster fitting is depending on the dimension of the network used.

**Review by Andrei32Ionescu — APPROVED** [review]

File: src/laplace.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (adelinacazacu) [reply src/laplace.jl:164]: [Here is a source](https://stats.stackexchange.com/questions/217736/is-there-a-3d-neural-network-and-how-to-code-it-in-r) that Mark sent to me in regards to this. It appears that in the case of Convolutional Neural Networks, the weights are still represented in the form of a matrix, so we cannot really have 3d+coordinates.

**Comment by MarkArdman** [comment]
There has been some discussion/debate surrounding the possibility of nodes being organized in forms [other](https://i.stack.imgur.com/U3iLJ.png) than [1 dimensional layers](https://res.cloudinary.com/edlitera/image/upload/c_fill,f_auto/v1660296216/blog/kbgm7toc4isbqsc7fwxb). We have come up with algorithms for both scenarios. It is difficult to find a definitive answer as to how neural networks are organized, but the general [convention](https://stats.stackexchange.com/questions/217736/is-there-a-3d-neural-network-and-how-to-code-it-in-r) seems to suggest that the latter image is the norm. We will contact our coach and client for clarifications and further information, and if necessary create a new issue for the generalized algorithm. Until further notice we have elected the current implementation to be satisfactory.
